fv/etcd_restart_test.go

```
@@ -0,0 +1,147 @@
+		defaultProfile.Spec.LabelsToApply = map[string]string{"default": ""}
+		defaultProfile.Spec.Egress = []api.Rule{{Action: api.Allow}}
+		defaultProfile.Spec.Ingress = []api.Rule{{Action: api.Allow}}
+		_, err := client.Profiles().Create(utils.Ctx, defaultProfile, utils.NoOptions)
+		Expect(err).NotTo(HaveOccurred())
+
+		// Wait until the tunl0 device appears; it is created when felix inserts the ipip module
+		// into the kernel.
+		Eventually(func() error {
+			links, err := netlink.LinkList()
+			if err != nil {
+				return err
+			}
+			for _, link := range links {
+				if link.Attrs().Name == "tunl0" {
+					return nil
+				}
+			}
+			return errors.New("tunl0 wasn't auto-created")
+		}).Should(BeNil())
+
+		// Create workloads, using that profile.  One on each "host".
+		for ii := range w {
+			wIP := fmt.Sprintf("10.65.%d.2", ii)
+			wIface := fmt.Sprintf("cali1%d", ii)
+			wName := fmt.Sprintf("w%d", ii)
+			w[ii] = workload.Run(felixes[ii], wName, wIface, wIP, "8055", "tcp")
+			w[ii].Configure(client)
+
+			hostW[ii] = workload.Run(felixes[ii], fmt.Sprintf("host%d", ii), "", felixes[ii].IP, "8055", "tcp")
+		}
+
+		cc = &workload.ConnectivityChecker{}
+	})
+
+	AfterEach(func() {
+		if CurrentGinkgoTestDescription().Failed {
+			for _, felix := range felixes {
+				felix.Exec("iptables-save", "-c")
+				felix.Exec("ipset", "list")
+				felix.Exec("ip", "r")
+			}
+		}
+
+		for _, wl := range w {
+			wl.Stop()
+		}
+		for _, wl := range hostW {
+			wl.Stop()
+		}
+		for _, felix := range felixes {
+			felix.Stop()
+		}
+
+		if CurrentGinkgoTestDescription().Failed {
+			etcd.Exec("etcdctl", "ls", "--recursive", "/")
+		}
+		etcd.Stop()
+	})
+
+	It("should survive an ungraceful etcd termination", func() {
+		By("having workload to workload connectivity", func() {
+			cc.ExpectSome(w[0], w[1])
+			cc.ExpectSome(w[1], w[0])
+			cc.CheckConnectivity()
+		})
+
+		By("killing etcd", func() {
+			// Kill etcd so that it cannot gracefully terminate connections.
```

I think, even a `kill -9` will gracefully terminate connections because the connection is socket gets cleaned up by the kernel.

One option is to nuke the conntrack entry inside the felix container.  I think you can delete by destination IP, and you have etcd's IP in-hand as etcd.IP (IIRC).  Another is to sneak in a raw iptables rule inside one of the containers, that lets you silently drop packets.

---

It's a good idea to check the graceful restart case too, of course!